akka-stream/src/main/scala/akka/stream/scaladsl/FlowWithContext.scala

```diff
@@ -0,0 +1,172 @@
+  def map[Ctx2](f: Ctx ⇒ Ctx2): Op[Ctx2] = OneToMany(f(context))
+  def flatMap[Ctx2](f: Ctx ⇒ Op[Ctx2]): Op[Ctx2] = f(context)
+}
+
+/**
+ * Describes a many-to-one type operator, which passes the contexts
+ * from many input elements to one output element.
+ */
+case class ManyToOne[+Ctx](contexts: Seq[Ctx]) extends Op[Ctx] {
+  def map[Ctx2](f: Ctx ⇒ Ctx2): Op[Ctx2] = ManyToOne(contexts.map(f))
+  def flatMap[Ctx2](f: Ctx ⇒ Op[Ctx2]): Op[Ctx2] = Op.manyToOne(contexts.map(f))
+}
+
+object Op {
+
+  /**
+   * Applies a many-to-one operation.
+   */
+  private[akka] def manyToOne[Ctx](contexts: Seq[Op[Ctx]]): ManyToOne[Ctx] = ManyToOne(contexts.flatMap {
+    case OneToOne(ctx)     ⇒ Seq(ctx)
+    case OneToMany(ctx, _) ⇒ Seq(ctx)
+    case ManyToOne(ctxs)   ⇒ ctxs
+    case NoOp              ⇒ Seq.empty[Ctx]
+  })
+
+  /**
+   * Applies a one-to-many operation.
+   */
+  private[akka] def oneToMany[Ctx](op: Op[Ctx]) = op match {
+    case OneToOne(ctx)       ⇒ OneToMany(ctx)
+    case otm: OneToMany[Ctx] ⇒ otm
+    case mtm: ManyToOne[Ctx] ⇒ mtm
+    case NoOp                ⇒ NoOp
+  }
+}
+
+trait FlowWithContextOps[+Ctx, +Out, +Mat] {
+  type Repr[+C, +O] <: FlowWithContextOps[C, O, Mat] {
+    type Repr[+CC, +OO] = FlowWithContextOps.this.Repr[CC, OO]
+    type Prov[+CC, +OO] = FlowWithContextOps.this.Prov[CC, OO]
+    type Under[+OO] = FlowWithContextOps.this.Under[OO]
+  }
+
+  type Prov[+C, +O] <: FlowOpsMat[(O, Op[C]), Mat]
+  type Under[+O] <: FlowOpsMat[O, Mat]
+
+  def via[Ctx2, Out2, Mat2](flow: Graph[FlowShape[(Out, Op[Ctx]), (Out2, Op[Ctx2])], Mat2]): Repr[Ctx2, Out2]
+
+  def map[Out2](f: Out ⇒ Out2): Repr[Ctx, Out2] =
+    via(flow.map { case (e, ctx) ⇒ (f(e), ctx) })
+
+  def mapAsync[Out2](parallelism: Int)(f: Out ⇒ Future[Out2]): Repr[Ctx, Out2] =
+    via(flow.mapAsync(parallelism) { case (e, ctx) ⇒ f(e).map(o ⇒ (o, ctx))(ExecutionContexts.sameThreadExecutionContext) })
+
+  def collect[Out2](f: PartialFunction[Out, Out2]): Repr[Ctx, Out2] =
+    via(flow.collect {
+      case (e, ctx) if f.isDefinedAt(e) ⇒ (f(e), ctx)
+    })
+
+  def filter(pred: Out ⇒ Boolean): Repr[Ctx, Out] =
+    collect { case e if pred(e) ⇒ e }
+
+  def filterNot(pred: Out ⇒ Boolean): Repr[Ctx, Out] =
+    collect { case e if !pred(e) ⇒ e }
+
+  def grouped(n: Int): Repr[Ctx, immutable.Seq[Out]] =
+    via(flow.grouped(n).map { elsWithContext ⇒
+      val (els, ctxs) = elsWithContext.unzip
+      (els, Op.manyToOne(ctxs))
+    })
+
+  def mapConcat[Out2](f: Out ⇒ immutable.Iterable[Out2]): Repr[Ctx, Out2] = statefulMapConcat(() ⇒ f)
+
+  def statefulMapConcat[Out2](f: () ⇒ Out ⇒ immutable.Iterable[Out2]): Repr[Ctx, Out2] = {
+    val fCtx: () ⇒ ((Out, Op[Ctx])) ⇒ immutable.Iterable[(Out2, Op[Ctx])] = { () ⇒ elWithContext ⇒
+      val (el, ctx) = elWithContext
+      val newCtx = Op.oneToMany(ctx)
+      f()(el).map(o ⇒ (o, newCtx))
+    }
+    via(flow.statefulMapConcat(fCtx))
+  }
+
```

Comment:
So it means that we have to define all operators that should support context propagation here. What is the fallback if a user need an operator that we have not defined here, e.g. `throttle`? Is it to convert it to a tuple stream and then back again?

Reply:
Sorry Patrik, misse this comment. Yes, you would go back.

Reply:
doing it with `via` might make that conversion not-too-painful perhaps?